test/powershell/engine/Basic/GroupPolicySettings.Tests.ps1

```diff
@@ -0,0 +1,257 @@
+        }
+
+        AfterEach {
+            Remove-item $KeyRoot -Recurse -Force
+            Remove-item $WinPSKeyRoot -Recurse -Force
+        }
+
+        It 'Execution policy test'{
+            function TestFeature
+            {
+                param([string]$KeyPath)
+
+                Set-ItemProperty -Path $KeyPath -Name EnableScripts -Value 1 -Force    
+
+                Set-ItemProperty -Path $KeyPath -Name ExecutionPolicy -Value 'Unrestricted' -Force
+                (Get-ExecutionPolicy) | Should -Be 'Unrestricted'
+                Set-ItemProperty -Path $KeyPath -Name ExecutionPolicy -Value 'AllSigned' -Force
+                (Get-ExecutionPolicy) | Should -Be 'AllSigned'
+                Set-ItemProperty -Path $KeyPath -Name ExecutionPolicy -Value 'RemoteSigned' -Force
+                (Get-ExecutionPolicy) | Should -Be 'RemoteSigned'
+
+                Remove-ItemProperty -Path $KeyPath -Name ExecutionPolicy -Force
+            }
+
+            TestFeature -KeyPath $KeyRoot
+
+            Set-ItemProperty -Path $KeyRoot -Name UseWindowsPowerShellPolicySetting -Value 1 -Force
+            TestFeature -KeyPath $WinPSKeyRoot
+        }
+
+        It 'Module logging policy test'{
+            function TestFeature
+            {
+                param([string]$KeyPath)
+            
+                $ModuleToLog = 'Microsoft.PowerShell.Utility'
+                $ModuleNamesKeyPath = Join-Path $KeyPath 'ModuleNames'
+                if (-not (Test-Path $ModuleNamesKeyPath)) {$null = New-Item $ModuleNamesKeyPath}
+            
+                Remove-Module $ModuleToLog -ErrorAction SilentlyContinue
+                Import-Module $ModuleToLog
+                (Get-Module $ModuleToLog).LogPipelineExecutionDetails | Should -Be $False # without GP logging for the module should be OFF
+            
+                # enable GP    
+                [string]$RareCommand = Get-Random
+                Set-ItemProperty -Path $KeyPath -Name EnableModuleLogging -Value 1 -Force
+                Set-ItemProperty -Path $ModuleNamesKeyPath -Name $ModuleToLog -Value $ModuleToLog -Force
+            
+                Remove-Module $ModuleToLog -ErrorAction SilentlyContinue
+                Import-Module $ModuleToLog # this will read and start using GP setting
+                (Get-Module $ModuleToLog).LogPipelineExecutionDetails | Should -Be $True # with GP logging for the module should be ON
+                
+                Get-Alias $RareCommand -ErrorAction SilentlyContinue | Out-Null
+            
+                (Get-Module $ModuleToLog).LogPipelineExecutionDetails = $False # turn off logging
+                Remove-ItemProperty -Path $KeyPath -Name EnableModuleLogging -Force # turn off GP setting
+                Remove-item $ModuleNamesKeyPath -Recurse -Force
+                # usually event becomes visible in the log after ~500 ms
+                For ($i=0; $i -le 50; $i++) # set timeout for 5 seconds
```

Comment:
We already have Wait-Until helper function.

Reply:
Good to know; Updated.
Also, looking at `Wait-UntilTrue` @JamesWTruher raised a good point that it will get stuck if the scriptblock has a blocking call (because it uses `$sb.Invoke()`) and the timeout won't fire. Probably it should be rewritten in async way using Jobs or something...